### ISSUE TITLE
BUG: scipy.special.cython_special.erf_inv called from numba broke in scipy-1.10

### ISSUE THREAD

**Issue by HDembinski** [scipy/scipy#18278]
### Describe your issue.

tl;dr: Something broke in scipy.special.cython_special.erf_inv in the last update from scipy-1.9 to scipy-1.10. When called from a numba.jitted function, it now returns always something which is approximately zero, e.g. 0 or 1.497547e-314, etc. It worked correctly before.

I am maintainting a library called [numba-stats](https://github.com/HDembinski/numba-stats), which provides a collection of numba-accelerated statistical functions from scipy.stats. The numba-stats versions are sometimes orders of magnitude faster than the scipy implementations, sometimes even though both call compiled routines like `ndtri`.

The problem is independent of the target platform, I see this bug on Linux, Macos, and Windows.

As a workaround, I am internally switching to using `ndtri` now, which works correctly and even makes my implementation simpler. I nevertheless wanted to report the issue.

### Reproducing Code Example

```python
# numba currently does not support scipy, so we cannot access
# scipy.stats.norm.ppf and scipy.stats.poisson.cdf in a JIT'ed
# function. As a workaround, we wrap special functions from
# scipy to implement the needed functions here.
from numba.extending import get_cython_function_address
from numba.types import WrapperAddressProtocol, float64
import scipy.special.cython_special as cysp


def get(name, signature):
    # create new function object with correct signature that numba can call by extracting
    # function pointer from scipy.special.cython_special; uses scipy/cython internals
    index = 1 if signature.return_type is float64 else 0
    pyx_fuse_name = f"__pyx_fuse_{index}{name}"
    if pyx_fuse_name in cysp.__pyx_capi__:
        name = pyx_fuse_name
    addr = get_cython_function_address("scipy.special.cython_special", name)

    # dynamically create type that inherits from WrapperAddressProtocol
    cls = type(
        name,
        (WrapperAddressProtocol,),
        {"__wrapper_address__": lambda self: addr, "signature": lambda self: signature},
    )
    return cls()

erfinv = get("erfinv", float64(float64))

# not bugged, just added to show that this technique works in principle
ndtri = get("ndtri", float64(float64))

from numba import njit


@njit
def testcase1(x):
    return erfinv(x)


@njit
def testcase2(x):
    return ndtri(x)


print(testcase1(0), testcase1(0.5), testcase1(1.0))
# 0.0 0.0 0.0
print(testcase2(0), testcase2(0.5), testcase2(1.0))
# -inf 0.0 inf
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
1.10.1 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: cmake
    found: true
    include directory: unknown
    lib directory: unknown
    name: OpenBLAS
    openblas configuration: unknown
    pc file directory: unknown
    version: 0.3.18
  lapack:
    detection method: cmake
    found: true
    include directory: unknown
    lib directory: unknown
    name: OpenBLAS
    openblas configuration: unknown
    pc file directory: unknown
    version: 0.3.18
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 12.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 12.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.33
  fortran:
    commands: gfortran
    linker: ld64
    name: gcc
    version: 4.9.0
  pythran:
    include directory: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-wiijwvrt/overlay/lib/python3.10/site-packages/pythran
    version: 0.12.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
Python Information:
  path: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-run-oo9bbb8v/cp310-macosx_x86_64/build/venv/bin/python
  version: '3.10'
```

**Comment by HDembinski** [comment]
Looking closer, I think I found the bug.
```py
from scipy.special import cython_special as cysp

print(cysp.__pyx_capi__["__pyx_fuse_0erfinv"])
# <capsule object "double (double, int __pyx_skip_dispatch)" at 0x10f032310>

print(cysp.__pyx_capi__["__pyx_fuse_1erfinv"])
# <capsule object "float (float, int __pyx_skip_dispatch)" at 0x10f032340>
```
My code assumes that "__pyx_fuse_1erfinv" is the version that uses double, and "__pyx_fuse_0erfinv" is the one for float, but this was swapped in scipy-1.10.

Since my code depends on implementation details of scipy, I assume that the error is on my end and I need to make sure that I select the correct overload. Feel free to close.

**Comment by ev-br** [comment]
Out of curiosity: has this ordering change in the scipy version or is it python version dependent, or maybe cython version dependent?
(Agreed it's likely a non-public implementation detail).

**Comment by HDembinski** [comment]
My experiments show that it depends on the scipy version only, Python version, platform etc. do not seem to be relevant.

**Comment by HDembinski** [comment]
> @HDembinski erfinv is a ufunc, and Numba should have support for those.

In my understanding ufuncs do not automatically work in Numba (and I just tried it out...). To call a function in Numba, I need a function pointer to the underlying compiled implementation, which I cannot get from the ufunc (to my knowledge).

**Comment by ev-br** [comment]
From 30000 feet, IIRC improved numpy/numba compatibility is under discussion for numpy 2.0. I'm sure if numpy does something in that directions (kermel pointers from ufuncs?), scipy ufunc machinery could follow. 
So maybe join that effort if you've cycles to spare?

**Comment by rgommers** [comment]
What I had in mind was, from https://numba.readthedocs.io/en/stable/reference/numpysupported.html:

_Numba understands calls to NumPy [ufuncs](http://docs.scipy.org/doc/numpy/reference/ufuncs.html) and is able to generate equivalent native code for many of them._

I had assumed that that worked for non-numpy ufuncs generated with the numpy ufunc machinery too. If this does not actually work for `scipy.special`, then I think making that work in Numba natively is the way to go. Either by making it aware of `scipy.special` specifically, or non-numpy ufuncs more generally.

**Comment by HDembinski** [comment]
> Numba understands calls to NumPy [ufuncs](http://docs.scipy.org/doc/numpy/reference/ufuncs.html) and is able to generate equivalent native code for many of them.

To my knowledge, support for each numpy ufunc has to be added by hand by the Numba devs using a bit of binding/emulation code, it is not automatic. See
https://github.com/numba/numba/blob/main/numba/np/ufunc_db.py
